SystemD Support--Limited

SChapa22 opened the pull request:
See the main() function in systemd.c for an example of use.

SChapa22 committed:
Delete WorkingProject directory

SChapa22 committed:
SupportSystemD

Three files, implementing limited support for use of SystemD and python bindings to those functions.
cgroup_create_scope_and_slice and cgroup_create_scope_user_slice create a scope in a custom slice or the user slice, respectively, and allow the caller to control delegation of scope (a yes/no choice) as well as unit mode (via defined values). Both will leave a permanent sleep process in the created scope, and return either an error code (a negative number, or 0) or the PID (positive) of the permasleep process.
cgroup_is_delegated determines if the given scope/slice is delegated, taking the full path to the desired SystemD unit.

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+#include <unistd.h>
+#include <systemd/sd-bus.h>
+
+#define CG_SYSTEMD_USER_SLICE_NAME "user.slice"
```
This macro isn't ever used.  Do you have a future plan for it?

SChapa22 replied:
Fixed, it is now used in create_scope_user_slice. I had also thought this (and the system.slice) should be exposed to the user--safer than letting them type it out themselves, but doesn't really need an enum if you'd rather remove it altogether.

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+
+//@todo Improve error messages for create_scope_and_slice & create_scope_user_slice; look at cg_delegate.py
+
+int cgroup_create_scope_and_slice(char *scope_name, char *slice_name, int delegated, char* mode){
```
Since the modes are declared above, they won't be visible to the user invoking the library.

Also, it's often more friendly (and more efficient) to use an enum as the mode parameter.

SChapa22 replied:
Fixed, though it doesn't look elegant: there's an enum for the user to use, and that gets mapped to an array of strings to get the true mode.

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	sd_bus *bus = NULL;
+	int ret = 0;
+	
+	int childPID = fork();
```
nitpick - in libcgroup world, we would use `child_pid`

SChapa22 replied:
Fixed

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	
+	int childPID = fork();
+
+	if(childPID < 0){ //No child process created
```
another nitpick but worth mentioning because other projects will get really upset by this.  I would write it as:
`if (childPID < 0) {`

SChapa22 replied:
Fixed

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	if(childPID < 0){ //No child process created
+		fprintf(stderr, "fork() failed: %d:%s\n", childPID, strerror(-childPID));
+	}
+	else if (childPID == 0){ //This is the child process, sleep forever
```
I know this is "work in progress" code... in the final code we don't want comments with `//`

SChapa22 replied:
Removed, excepting the main() which will be removed when all other parts are ready for integration.

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+			"/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager",
+			 "StartTransientUnit");
+		if (ret<0){
+			fprintf(stderr, "sd_bus_message_new_method_call failed: %d:%s\n", ret, strerror(-ret));
```
Again, `fprintf()` isn't libcgroup-friendly.  libcgroup uses `cgroup_err()` in these cases.

SChapa22 replied:
I have an attempted fix (including libcgroup-internal.h and tools/tools-common.h) that's giving me the following error message:
`systemd.c:11:10: fatal error: libcgroup-internal.h: No such file or directory
   11 | #include <libcgroup-internal.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
`

SChapa22 replied:
Fixed, now that it's inside libcgroup I can access the -internal.h and compile with cgroup_err

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+			goto out;
+		}
+
+		ret = sd_bus_message_read(reply, "o", &object);
```
Do we need to read the reply since we're not doing anything with it?

SChapa22 replied:
I don't think so, so this has been removed. If the intent was to force the message to process before returning, simply the fact that object has been assigned in the previous call would do that.

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+		}
+		
+		// Parent thread will return PID of child thread
+		ret = childPID;
```
This feels weird to me.  Normally we return an error code (or 0) in the return int.

SChapa22 replied:
The caller should be told the PID of the permanent sleeper process (or else given some way to find it), else the caller has no clue which process to kill (and may kill itself--child and parent have the same name!). While it isn't necessary to save the user the mishap, isn't it safer to do so?

drakenclimber replied:
Yes, sharing the PID with the caller would be nice, but overloading a return call is dangerous.  It would probably be better to have it as an out parameter in the function call.  Something like this:

```
 int cgroup_create_scope_and_slice(char *scope_name, char *slice_name, int delegated, char* mode, pid_t * const placeholder_pid){ 
 ```
 
The const is optional, but I'm a big fan of `const` :)

SChapa22 replied:
Fixed, with const

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+		
+		ret = sd_bus_open_system(&bus);
+		if (ret<0){
+			fprintf(stderr, "sd_bus_open_system failed: %d:%s\n", ret, strerror(-ret));
```
FYI, 99% of the libcgroup functions use `ECG*` return codes rather than `-EINVAL`, etc.  We shouldn't deviate from this paradigm unless we have a really good reason.

SChapa22 replied:
Postponed until I can get cgerr() calls working. Can I just return the number of the ecg error code, or is there also a function to return the string?

SChapa22 replied:
Postponed until I can get cgerr() calls working. Can I just return the number of the ecg error code, or is there also a function to return the string?

drakenclimber replied:
Look at examples in libcgroup.  Here's one in api.c that seems reasonable
https://github.com/libcgroup/libcgroup/blob/main/src/api.c#L1823

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	}
+
+	//@todo Does object need to be freed?
+	out:
```
`out:` should not be indented.

SChapa22 replied:
Fixed

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	return ret;
+}
+
+int cgroup_create_scope_user_slice(char *scope_name, int delegated, char *mode){
```
Do we really need a dedicated function to create the scope in the user.slice?

And more importantly, if we do need it, why all of the code duplication?  Wouldn't this work:

```
int cgroup_create_scope_user_slice(char *scope_name, int delegated, char *mode)
{
    return cgroup_create_scope_and_slice(scope_name, USER_SLICE, delegated, mode);
}
```

SChapa22 replied:
Fixed, with present code left in block comment. As I said before, my original attempt to do this had sporadic errors crop up when getxattr was called after the wrapper function, specifically error 1-Operation Invalid (break happened on what is now line 235). I couldn't pin down what caused the error so as to consistently trigger it, but repeated runs revealed it was there, so I went with duplicate code over potentially error-generating code. This hasn't recurred since re-implementing the wrapper, but I figured I'd note it was there at some point in case it comes back.

SChapa22 replied:
Commented code removed.

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+
+//@todo Improve error messages for create_scope_and_slice & create_scope_user_slice; look at cg_delegate.py
+
+int cgroup_create_scope_and_slice(char *scope_name, char *slice_name, int delegated, char* mode){
```
What if the slice already exists?  Then what happens?

SChapa22 replied:
Then there will be an error message regarding the creation of the slice, but the desired scope will be created in that slice. We don't have a GOOD error message, though, and I'm not sure how to fix that from this file (I suspect an error has to be defined elsewhere, then thrown from here).
<img width="484" alt="AlreadyExistsError" src="https://user-images.githubusercontent.com/99231232/179862117-dac847f6-8569-4d59-af30-d1d6bd069fe5.png">

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	return ret;
+}
+
+int cgroup_is_delegated(char *path){
```
Again, I'm not fond of forcing the user to build up the path.  This will lead to many problems

SChapa22 replied:
Will work on finding path based on a .scope/.slice string tomorrow. As of now, still pending.

SChapa22 replied:
Fixed

SChapa22 replied:
Un-fixed: need to decide how to find slice/scope, since originally implemented fix (tree-walk) is not ideal

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+		fprintf(stderr, "getxattr failed: %d:%s\n", ret, strerror(-ret));
+		fflush(stderr);
+	}
+	else if (strcmp(result, "1")==0){
```
Static analysis tools flag `strcmp()` as dangerous.  `strncmp()` is preferred.

SChapa22 replied:
Fixed

drakenclimber commented on src/systemd.c:
```diff
@@ -0,0 +1,286 @@
+	return ret;
+}
+
+int main (int argc, char* argv[]){
```
`main()` needs to go away and we need to integrate this code into the libcgroup library

SChapa22 replied:
True, but until all other issues are resolved, I'm keeping it to more easily test all this. When you tell me everything else is perfect, I'll happily remove it.

SChapa22 replied:
Removed, testing is now being done in SC_test.c

drakenclimber commented on src/systemd.h:
```diff
@@ -0,0 +1,32 @@
+#define __LIBC_SYSTEMD
+
+/**
+ * Function to create a scope of specified name within the specified slice; the scope is delegated if
```
Look at my comment in the makefile patch I shared with you.  Libcgroup uses doxygen-style comments in the *.h files.
https://github.com/drakenclimber/libcgroup/commit/98e8f16230048831e7ba30f0092ff24148a3076b#diff-874ad12789566e63ecf1207b33dcf475df61d3036be3ef7fc9f91cac883399b2R15

SChapa22 replied:
Fixed; header and footer also copied over so the naming of the ifndef and comments surrounding those should now be correct.

drakenclimber commented on src/systemd.py:
```diff
@@ -0,0 +1,13 @@
+# Equivalent to the defined variables in systemd.c
```
Let's ignore the python stuff until we get the C code nailed down better.

SChapa22 replied:
Agreed

SChapa22 committed:
RevisionsV1

See Comments from Pull Request 1

SChapa22 committed:
RevisionsV2

Erroneously uploaded the non-current file for RevisionsV1. See comments on pull request 1 for details.

SChapa22 commented:
Latest commit of WorkInProgress branch now addresses above comments.

SChapa22 committed:
FindScopeSlice

Added functionality to find a scope or slice by name rather than by full path.
